drivers/encx24j600/encx24j600.c

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     driver_encx24j600
+ * @{
+ *
+ * @file
+ * @brief       Internal functions for the ENCX24J600 driver
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include "mutex.h"
+#include "encx24j600.h"
+#include "encx24j600_internal.h"
+#include "encx24j600_defines.h"
+#include "vtimer.h"
+
+#include "net/dev_eth.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define ENCX24J600_SPI_SPEED SPI_SPEED_1MHZ
+
+#define ENCX24J600_INIT_DELAY 100000U
+
+#define ENC_BUFFER_START 0x0000
+#define ENC_BUFFER_SIZE  0x6000
+#define ENC_BUFFER_END   0x5FFF
+#define RX_BUFFER_START (0x5340) /* Default value */
+#define RX_BUFFER_END   (ENC_BUFFER_END)
+#define TX_BUFFER_LEN   (0x2000)
+#define TX_BUFFER_END   (RX_BUFFER_START)
+#define TX_BUFFER_START (TX_BUFFER_END - TX_BUFFER_LEN)
+
+static void receive_packets(encx24j600_t *dev);
+static void cmd(encx24j600_t *dev, char cmd);
+static void reg_set(encx24j600_t *dev, uint8_t reg, uint16_t value);
+static uint16_t reg_get(encx24j600_t *dev, uint8_t reg);
+static void reg_clear_bits(encx24j600_t *dev, uint8_t reg, uint16_t mask);
+
+/* dev_eth interface */
+static void _isr(dev_eth_t *dev);
+static int _init(dev_eth_t *dev);
+static int _send(dev_eth_t *dev, char *pkt, int len);
+static void _get_mac_addr(dev_eth_t *dev, uint8_t* buf);
+static int _recv(dev_eth_t *encdev, char* buf, int len);
+
+static inline int _get_promiscous(dev_eth_t *dev) {
+    return 1;
+}
+
+static inline int _set_promiscous(dev_eth_t *dev, int value) {
+    return 1;
+}
+
+const static eth_driver_t eth_driver_encx24j600 = {
+    .init = _init,
+    .send = _send,
+    .recv = _recv,
+    .get_mac_addr = _get_mac_addr,
+    .get_promiscous = _get_promiscous,
+    .set_promiscous = _set_promiscous,
+    .isr = _isr,
+};
+
+static inline void lock(encx24j600_t *dev) {
```

Comment:
actually, it is not quite clear to me, why this device driver level locking is needed. The lower level SPI bus is protected by it's own lock, and the usage model of the network interface drivers so far is that they are only used inside a single thread (e.g. by a single user), so that device driver level locking is not really needed. Seems I am missing something?!